pkg/controller/networkpolicy/networkpolicy_controller.go

```diff
@@ -1083,29 +1084,53 @@ func (n *NetworkPolicyController) syncAddressGroup(key string) error {
 	return nil
 }
 
-func (n *NetworkPolicyController) populateAddressGroupMemberSet(g *antreatypes.AddressGroup) controlplane.GroupMemberSet {
-	var pods []*v1.Pod
-	var externalEntities []*v1alpha2.ExternalEntity
+// getAddressGroupMemberSet knows how to construct a GroupMemberSet that contains
+// all the entities selected by an AddressGroup.
+func (n *NetworkPolicyController) getAddressGroupMemberSet(g *antreatypes.AddressGroup) controlplane.GroupMemberSet {
 	// Check if an internal Group object exists corresponding to this AddressGroup.
-	_, found, _ := n.internalGroupStore.Get(g.Name)
+	groupObj, found, _ := n.internalGroupStore.Get(g.Name)
 	if found {
 		// This AddressGroup is derived from a ClusterGroup.
-		pods, externalEntities = n.groupingInterface.GetEntities(clusterGroupType, g.Name)
-	} else {
-		pods, externalEntities = n.groupingInterface.GetEntities(addressGroupType, g.Name)
+		group := groupObj.(*antreatypes.Group)
+		return n.getClusterGroupMemberSet(group)
+	}
+	return n.getMemberSetForGroupType(addressGroupType, g.Name)
+}
+
+// getClusterGroupMemberSet knows how to construct a GroupMemberSet that contains
+// all the entities selected by a ClusterGroup. For ClusterGroup that has childGroups,
+// the members are computed as the union of all its childGroup's members.
+// This function assumes that there's no loop in Group reference.
+func (n *NetworkPolicyController) getClusterGroupMemberSet(group *antreatypes.Group) controlplane.GroupMemberSet {
+	if len(group.ChildGroups) == 0 {
+		return n.getMemberSetForGroupType(clusterGroupType, group.Name)
+	}
+	groupMemberSet := controlplane.GroupMemberSet{}
+	for _, childName := range group.ChildGroups {
+		childGroup, found, _ := n.internalGroupStore.Get(childName)
+		if found {
+			child := childGroup.(*antreatypes.Group)
+			groupMemberSet = groupMemberSet.Union(n.getClusterGroupMemberSet(child))
```

Comment:
ditto, why didn't it call getMemberSetForGroupType directly?

Reply:
Replied on the other thread